server/src/main/kotlin/com/kamelia/jellyfish/rest/file/Service.kt

```diff
@@ -1,79 +1,71 @@
-            NIOFiles.createDirectories(directory)
-            val filePath = directory.resolve(fileCode)
-            NIOFiles.write(filePath, fileBytes)
-            val fileMimeType = URLConnection.guessContentTypeFromName(fileName) ?: "application/octet-stream"
-            val fileSize = NIOFiles.size(filePath)
+        if (fileName.isBlank()) return QueryResult.badRequest("errors.file.name.empty")
 
-            QueryResult.ok(
-                Files.create(
-                    code = fileCode,
-                    name = fileName,
-                    mimeType = fileMimeType,
-                    size = fileSize,
-                    creator = creator
-                ).toRepresentationDTO()
-            )
-        }
+        val (code, type, size) = FileUtils.write(creator.uuid, part)
+        return QueryResult.ok(
+            Files.create(
+                code = code,
+                name = fileName,
+                mimeType = type,
+                size = size,
+                creator = creator
+            ).toRepresentationDTO()
+        )
     }
 
-    private suspend fun generateUniqueCode(): String {
-        repeat(10) {
-            val code = String.random(10)
-            if (Files.findByCode(code) == null) {
-                return code
+    suspend fun getFile(
+        code: String,
+        user: User?,
+    ): QueryResult<FileRepresentationDTO, ErrorDTO> = Files
+        .findByCode(code)
+        ?.takeUnless { file ->
+            val isPrivate = file.visibility == FileVisibility.PRIVATE
+            val notHasPermission = user?.let {
+                it.role lt UserRole.ADMIN && file.ownerId != it.uuid
             }
+            isPrivate && (notHasPermission ?: true)
+        }?.let { file ->
+            QueryResult.ok(file.toRepresentationDTO())
         }
-        throw UploadCodeGenerationException()
-    }
+        ?: QueryResult.notFound()
 
-    suspend fun getFiles(user: User, page: Long, pageSize: Int): QueryResult<FilePageDTO, List<ErrorDTO>> {
-        val files = user.getFiles(page, pageSize)
-        val total = user.countFiles()
-        return QueryResult.ok(FilePageDTO(
-            PageDTO(
-                files.map { it.toRepresentationDTO() },
-                page,
-                pageSize,
-                ceil(total / pageSize.toDouble()).toLong(),
-                total
+    suspend fun getFiles(
+        user: User,
+        page: Long,
+        pageSize: Int,
+        definition: PageDefinitionDTO,
+    ): QueryResult<FilePageDTO, List<ErrorDTO>> {
+        val (files, total) = user.getFiles(page, pageSize, definition)
+        return QueryResult.ok(
+            FilePageDTO(
+                PageDTO(
+                    files.map { it.toRepresentationDTO() },
+                    page,
+                    pageSize,
+                    ceil(total / pageSize.toDouble()).toLong(),
+                    total
+                )
             )
-        ))
+        )
     }
```

Comment:
I might be wrong, but with this method we never perform a permission check. I mean an admin could see private files. 

I think we should come up with a final convention of files visibility according to role. The owner should be able to see everything but an admin shouldn't. Let's keep a bit of privacy.

Reply:
Okay fair enough